Vungle/VungleRouter.m

```diff
@@ -528,44 +529,38 @@ - (void)clearDelegateWithState:(BannerRouterDelegateState)state placementID:(NSS
 
 - (void)clearWaitingList
 {
-    for (id key in self.waitingListDict) {
-        id<VungleRouterDelegate> delegateInstance = [self.waitingListDict objectForKey:key];
-        
-        NSString *targetPlacementID = [delegateInstance getPlacementID];
-        BOOL needRequestAd = YES;
-        
-        // If a placement is requested as priority placement at init call, no need to request again.
-        if ([self.prioritizedPlacementID isEqualToString:targetPlacementID]) {
-            needRequestAd = NO;
-        }
-        
-        if ([delegateInstance respondsToSelector:@selector(getBannerSize)]) {
-            NSString *id = [delegateInstance getPlacementID];
-            CGSize size = [delegateInstance getBannerSize];
-            [self requestBannerAdWithPlacementID:id size:size delegate:delegateInstance needRequestAd:needRequestAd];
-        } else {
-            if (![self.delegatesDict objectForKey:key]) {
-                [self.delegatesDict setObject:delegateInstance forKey:key];
-            }
-            
-            if (!needRequestAd) {
-                continue;
-            }
-            
-            NSError *error = nil;
-            if ([[VungleSDK sharedSDK] loadPlacementWithID:key error:&error]) {
-                MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", key);
+    @synchronized (self) {
+        for (id key in self.waitingListDict) {
+            id<VungleRouterDelegate> delegateInstance = [self.waitingListDict objectForKey:key];
+
+            if ([delegateInstance respondsToSelector:@selector(getBannerSize)]) {
+                NSString *id = [delegateInstance getPlacementID];
+                CGSize size = [delegateInstance getBannerSize];
+                [self requestBannerAdWithPlacementID:id size:size delegate:delegateInstance];
             } else {
-                if (error) {
-                    MPLogInfo(@"Vungle: Unable to load an ad for Placement ID :%@, Error %@", key, error);
+                if (![self.delegatesDict objectForKey:key]) {
+                    [self.delegatesDict setObject:delegateInstance forKey:key];
+                }
+
+                if ([self isAdAvailableForPlacementId:key]) {
+                    [delegateInstance vungleAdDidLoad];
+                    continue;
+                }
+
+                NSError *error = nil;
+                if ([[VungleSDK sharedSDK] loadPlacementWithID:key error:&error]) {
+                    MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", key);
+                } else {
+                    if (error) {
+                        MPLogInfo(@"Vungle: Unable to load an ad for Placement ID :%@, Error %@", key, error);
+                    }
+                    [delegateInstance vungleAdDidFailToLoad:error];
```

Comment:
can `error` be nil here?

Reply:
@crbinsf Got it, I will add the logic when error is nil. Thanks a lot for pointing it out. :)

Reply:
@crbinsf Added, thanks a lot. :)